### ISSUE TITLE
configurable storage for union type

### ISSUE THREAD

**Issue by kernel-io** [ash-project/ash#694]
Following on from our conversations on Discord, it would be helpful to be able to configure a simpler storage type for the union type.

Ideally we would just store the json directly in the database with a hidden type field (which is configurable with the `tag:` key, i.e: "__schema").

**Comment by zachdaniel** [comment]
I'm happy to implement this next week, as I think in some cases this can definitely be a much simpler strategy. However, I think this is also a relatively light-weight change, so PRs welcome! What we would want to add is something like a `storage_strategy`. To start with, we can add one additional strategy for what you're looking for. For example:

```elixir
types: [...],
storage_strategy: :explicit | :implicit # explicit is the current behavior, and would be the default
```

## Explicit
Everything is stored in a map w/ `{type: type, value: the_value}`

## Implicit
Everything is stored "as is", and is checked for its type on the way out. Tagged unions would get fast tracked and so would not have a significant performance degradation. By "fast tracked" I mean that we first check if the map matches any tagged unions, and if not then we try loading it as the remaining types in order. The implications of `implicit` are that if you have something that could be a `:string` or an `:integer` that it is very important to have them in the right order, i.e most specific to least specific. The simple answer for that would be to suggest that, when using the implicit strategy, everything in the union has a tag.

**Comment by kernel-io** [comment]
I think `implicit` would always have a tag / _should_ always have a tag.

**Comment by zachdaniel** [comment]
Agreed. It should be fine if they don't all have tags, just some limitations to accept. If they all have tags then there are no significant penalties.

**Comment by zachdaniel** [comment]
I might even consider making `implicit` the default in 3.0, as it has some benefits...

**Comment by zachdaniel** [comment]
Merged in main. I chose better names than `explicit` vs `implicit`. The options are `storage: :type_and_value | :map_with_tag`
